photoeditor/src/main/java/ja/burhanrashid52/photoeditor/DrawingView.kt

```diff
@@ -0,0 +1,217 @@
+
+    // eraser parameters
+    private var isErasing = false
+    var eraserSize = DEFAULT_ERASER_SIZE
+
+    // endregion
+    private fun createPaint(): Paint {
+        val paint = Paint()
+        paint.isAntiAlias = true
+        paint.isDither = true
+        paint.style = Paint.Style.STROKE
+        paint.strokeJoin = Paint.Join.ROUND
+        paint.strokeCap = Paint.Cap.ROUND
+        paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_OVER)
+
+        // apply shape builder parameters
+        currentShapeBuilder?.apply {
+            paint.strokeWidth = this.shapeSize
+            paint.alpha = this.shapeOpacity
+            paint.color = this.shapeColor
+        }
+
+        return paint
+    }
+
+    private fun createEraserPaint(): Paint {
+        val paint = createPaint()
+        paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.CLEAR)
+        return paint
+    }
+
+    private fun setupBrushDrawing() {
+        //Caution: This line is to disable hardware acceleration to make eraser feature work properly
+        setLayerType(LAYER_TYPE_HARDWARE, null)
+        visibility = GONE
+        currentShapeBuilder = ShapeBuilder()
+    }
+
+    fun clearAll() {
+        drawShapes.clear()
+        redoShapes.clear()
+        invalidate()
+    }
+
+    fun setBrushViewChangeListener(brushViewChangeListener: BrushViewChangeListener?) {
+        viewChangeListener = brushViewChangeListener
+    }
+
+    public override fun onDraw(canvas: Canvas) {
+        for (shape in drawShapes) {
+            shape?.shape?.draw(canvas, shape.paint)
+        }
+    }
+
+    /**
+     * Handle touch event to draw paint on canvas i.e brush drawing
+     *
+     * @param event points having touch info
+     * @return true if handling touch events
+     */
+    @SuppressLint("ClickableViewAccessibility")
+    override fun onTouchEvent(event: MotionEvent): Boolean {
+        return if (isDrawingEnabled) {
+            val touchX = event.x
+            val touchY = event.y
+            when (event.action) {
+                MotionEvent.ACTION_DOWN -> onTouchEventDown(touchX, touchY)
+                MotionEvent.ACTION_MOVE -> onTouchEventMove(touchX, touchY)
+                MotionEvent.ACTION_UP -> onTouchEventUp(touchX, touchY)
+            }
+            invalidate()
+            true
+        } else {
+            false
+        }
+    }
+
+    private fun onTouchEventDown(touchX: Float, touchY: Float) {
+        createShape()
+        currentShape?.shape?.startShape(touchX, touchY)
+    }
+
+    private fun onTouchEventMove(touchX: Float, touchY: Float) {
+        currentShape?.shape?.moveShape(touchX, touchY)
+    }
+
+    private fun onTouchEventUp(touchX: Float, touchY: Float) {
+        currentShape?.apply {
+            shape.stopShape()
+            endShape(touchX, touchY)
+        }
+    }
+
+    private fun createShape() {
+        val shape: AbstractShape
+        var paint = createPaint()
+        when {
```

Comment:
Can we use the when on the shapeType so that we don't need to explicitly in condition 
```
when(currentShapeBuilder?.shapeType){
   ShapeType.OVAL -> {}
}
```

Reply:
@burhanrashid52 so I may be reading it wrong, but would that mean we would need to move the `isErasing` condition out of the `when` block?

Reply:
Yes @lucianocheng

Reply:
So something like this to manage the `else` `BrushShape` condition? @burhanrashid52 

```
shape = BrushShape()
if (isErasing) {
  paint = createEraserPaint()
} else {
  when(currentShapeBuilder?.shapeType){
     ShapeType.OVAL -> {
         shape = OvalShape()
      }
  }
  ...
}
```

Reply:
Yes @lucianocheng

Reply:
👍 addressed.